src/coreclr/jit/gentree.cpp

```diff
@@ -15725,8 +15725,18 @@ GenTree* Compiler::gtNewTempAssign(
         // 2. The zero constant created by local assertion propagation.
         // In these cases, we can use the type of the merge return for the assignment.
         assert(val->gtEffectiveVal(true)->OperIs(GT_IND, GT_BLK, GT_LCL_FLD, GT_CNS_INT));
-        assert(tmp == genReturnLocal);
-        valStructHnd = lvaGetStruct(genReturnLocal);
+        if (tmp == genReturnLocal)
+        {
+            valStructHnd = lvaGetStruct(genReturnLocal);
+        }
+        else
+        {
+            // We're spilling IND<struct> to a temp from RETURN(IND<struct>) expression
+            // TODO-ADDR: delete once "IND<struct>" nodes are no more
+            assert(genReturnLocal == BAD_VAR_NUM);
+            genReturnLocal = tmp;
+            valStructHnd   = info.compMethodInfo->args.retTypeClass;
```

Comment:
This is an ugly workaround... I would recommend doing `lvaGrabTemp + lvaSetStruct` manually instead of this (`ReplaceWithLclVar` accepts a `tempNum` parameter for this).